extensions/src/tables/index.ts

```diff
@@ -0,0 +1,297 @@
+  MenuItem, SaveDataHandler
+} from "$common";
+
+type Details = {
+  name: "Tables",
+  description: "Make and use tables with rows and columns",
+  iconURL: "tables.png",
+  insetIconURL: "tables.svg"
+};
+
+type Blocks = {
+  createTable: ButtonBlock;
+  addTable: (name: string, rows: number, columns: number) => void;
+  removeTable: (table: string) => void;
+  insertColumn: (table: string) => void;
+  insertRow: (table: string) => void;
+  insertValueAt: (table: string, value: any, row: number, column: number) => void;
+  getValueAt: (table: string, row: number, column: number) => number;
+  numberOfRows: (table: string) => number;
+  numberOfColumns: (table: string) => number;
+  highestValueOfColumn: (table: string, column: number) => number;
+  highestValueOfRow: (table: string, row: number) => number;
+  indexOfHighestColumnValue: (table: string, column: number) => number;
+  indexOfHighestRowValue: (table: string, row: number) => number;
+  showTable: ButtonBlock;
+}
+
+export default class Tables extends Extension<Details, Blocks> {
+  tables: Record<string, number[][]>;
+  tableNamesArg: any;
+  defaultNumberArg: any;
+
+  // save tables to .sb3 when project is saved
+  saveDataHandler = new SaveDataHandler({
+    Extension: Tables,
+    onSave: (self) => { return self.tables },
+    onLoad: (self, tables) => { self.tables = tables },
+  });
+
+  init(env: Environment) {
+    if (!this.tables) {
+      this.tables = {};
+      this.tables.myTable = [];
+      this.tables.myTable.push([0]);
+    }
+    
+    // dynamic retriever of table names for block dropdowns
+    this.tableNamesArg = {
+      type: ArgumentType.String,
+      options: {
+        getItems: this.getTableNames.bind(this),
+        acceptsReporters: true,
+        handler: (reported: any) => {
+          if (this.getTableNames().indexOf(reported) === undefined) {
+            alert(`no table with name ${reported} exists`);
+            return 'myTable';
+          }
+          return reported;
+        }
+      }
+    };
+
+    this.defaultNumberArg = {
+      type: ArgumentType.Number,
+      defaultValue: 1
+    };
+  }
+
+  getTableNames(): MenuItem<string>[] {
+    return Object.keys(this.tables).map(
+      tableName => ({
+        text: tableName,
+        value: tableName
+      })
+    );
+  }
+
+
+  newTable(info: { name: string, rows: number, columns: number }) {
+    const { name, rows, columns } = info;
+    this.tables[name] = [];
+    for (let i = 0; i < rows; i++) {
+      let newRow = [];
+      for (let j = 0; j < columns; j++) newRow.push(0);
+      this.tables[name].push(newRow);
+    }
+  }
+
+  changeTableValue(info: { name: string, row: number, column: number, value: number }) {
+    const { name, row, column, value } = info;
+    this.tables[name][row][column] = value;
+  }
+
+  defineBlocks(): Tables["BlockDefinitions"] {
+    return {
+      // button that opens a modal to create a new table
```

Comment:
[NONBLOCKER / just FYI] Something kind of cool you could do is actually add a jsdoc-style comment to each block type up on line ~15. That would allow you to comment the params the function takes, but you'll still see that comment when hovering over of these definitions. For example (up where you declare the blocks type):

```ts
type Blocks = {
  /**
   * button that opens a modal to create a new table
   */
  createTable: ButtonBlock;
  ...
}
```

And when you hover over `createTable` here you should see your comment.